### PR TITLE
Breaking: Unordered, fast, collision resistant non-crypto hash

### PR DESCRIPTION
**Pull request by That3Percent**
This PR completely overhauls the APIs to be more organized and make sense. It also adds a multiset hash implementation for the fast variant, making it conceptually the same as the cryptographic variant (enabling the API cleanup) and making it a candidate for the PoI hash in graph-node.

### CODE REVIEW

File: src/crypto/hasher.rs [hunk 1]
```diff
@@ -1,114 +1,86 @@
...
 use num_traits::identities::One;
 use std::default::Default;
 
+// TODO: Consider using a Solinas prime
+/*
+From Jackson:
+    In particular we could change the prime to be a Solinas prime.
+    If we implement the algorithm for fast modular multiplication around a solinas prime then we get a big speed up.
+    So the changes would be just change the public parameter prime in the codebase, and don’t just naively multiply and reduce, but write the algorithm to take advantage of the structure inherent in Solinas primes
+    (they are prime numbers that have really low hamming weights, a sort of generalization of Mersenne primes — and so computers love these numbers)
```
Comment (lutter) [inline src/crypto/hasher.rs:15]: Is this still something we are considering?

File: src/crypto/hasher.rs [hunk 2]
```diff
@@ -1,114 +1,86 @@
...
+*/
 lazy_static! {
     static ref P: UBig = "50763434429823703141085322590076158163032399096130816327134180611270739679038131809123861970975131471260684737408234060876742190838745219274061025048845231234136148410311444604554192918702297959809128216170781389312847013812749872750274650041183009144583521632294518996531883338553737214586176414455965584933129379474747808392433032576309945590584603359054260866543918929486383805924215982747035136255123252119828736134723149397165643360162699752374292974151421555939481822911026769138419707577501643119472226283015793622652706604535623136902831581637275314074553942039263472515423713366344495524733341031029964603383".parse().unwrap();
```
Comment (lutter) [inline src/crypto/hasher.rs:18]: Where does this number come from? How do we know that it is a prime?

File: src/crypto/hasher.rs [hunk 3]
```diff
@@ -1,114 +1,86 @@
...
 }
 
 /// Based on https://crypto.stackexchange.com/a/54546
 ///
-/// The idea here is to use the SequenceNumber to unambiguously identify each
+/// The idea here is to use the FieldAddress to unambiguously identify each
 /// field as within it's own database cell, and use an online order-independent
 /// aggregator of the cells to produce a final result.
```
Comment (lutter) [inline src/crypto/hasher.rs:25]: I assume this means that a struct/record is turned into a set of pairs `(fieldAddr, value)` and their hash values are then hashed order-independently, i.e., as a set.

File: src/crypto/hasher.rs [hunk 4]
```diff
@@ -1,114 +1,86 @@
...
 ///
 /// Within this framework a huge struct can be hashed incrementally or even in
-/// parallel as long as sequence numbers are deterministically produced to
-/// identify parts within the struct. Conveniently, the SequenceNumber::skip
+/// parallel as long as field addresses are deterministically produced to
```
Comment (lutter) [inline src/crypto/hasher.rs:28]: What exactly does 'deterministically' mean here? I assume it means that the mapping from field name in the struct to field address is fixed and does not change as fields are reordered or renamed across code changes.

File: src/crypto/hasher.rs [hunk 5]
```diff
@@ -1,114 +1,86 @@
...
+/// uniquely identify parts within the struct. Conveniently, the FieldAddress::skip
 /// method can be used to jump to parts of a vec or struct efficiently.
-pub struct SetHasher {
+pub struct CryptoStableHasher {
     // TODO: (Performance). We want an int 2056 + 2048 = 4104 bit int (u4160 if using a word size of 64 at 65 words)
     // That's enough to handle any sequence of mixin operations without overflow.
```
Comment (lutter) [inline src/crypto/hasher.rs:33]: Is this still a TODO? What's the significance of 2056 + 2048 and why does that prevent overflow?

File: src/crypto/hasher.rs [hunk 6]
```diff
@@ -1,114 +1,86 @@
...
-        Default::default()
-    }
-    #[inline]
-    fn mixin(&mut self, digits: &UBig) {
-        profile_method!(mixin);
+    fn mixin_raw(&mut self, digits: &UBig) {
+        profile_method!(mixin_raw);
 
         self.value *= digits;
         self.value %= &*P;
```
Comment (lutter) [inline src/crypto/hasher.rs:51]: Would it make sense to use `ibig::modular` here in the hope that this modular multiplication of n bit numbers takes just a little more than n bits of scratch space instead of 2n? (I haven't looked how `ibig::modular` implements multiplication, so this might be moot)

File: src/crypto/hasher.rs [hunk 7]
```diff
@@ -120,4 +92,18 @@ impl StableHasher for SetHasher {
         hasher.update(&le);
         hasher.finalize().into()
     }
+
+    fn to_bytes(&self) -> Self::Bytes {
+        profile_method!(to_bytes);
+        self.value.to_le_bytes()
+    }
+    /// Panics if the bytes are not in a valid format.
```
Comment (lutter) [inline src/crypto/hasher.rs:100]: This should say 'Panics if `bytes` is longer than 257'. I don't know though where 257 comes from. Is `P` a 2056 bit prime? The comment in `write` seems to indicate that (assuming it should say '1 byte longer' instead of '1 bit longer')

File: src/fast/address.rs [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+    /// This test demonstrates that our choice of primes and algorithm is a good
+    /// one for our use case of common structures to be digested by trying every
+    /// permutation of all structs several deep and long Vecs for children and
+    /// asserting 0 collisions on over 11 million common <u64>
+    /// paths. Just for kicks I
+    /// ran it on over 700 million paths before committing, but this test did
+    /// not complete in a reasonable enough amount of time to be committed.
+    /// Larger than that and we get dangerously close to birthday collisions
+    /// anyway so I'm calling this good enough.
```
Comment (lutter) [inline src/fast/address.rs]: Very minor, but with 128 bit hash values, you need about 2*10^19 values for a probability of a collision to be >= 1/2

File: src/fast/u192.rs [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        let me = &self.0;
+        let you = &other.0;
+
+        let (r0, overflow) = me[0].overflowing_add(you[0]);
+        let (res, overflow1) = me[1].overflowing_add(you[1]);
+        let (r1, overflow2) = res.overflowing_add(overflow as u64);
+
+        let r2 = me[2]
+            .wrapping_add(you[2])
+            .wrapping_add(overflow1 as u64 + overflow2 as u64);
```
Comment (lutter) [inline src/fast/u192.rs]: I don't understand the math here: AFAICT, we want to compute 
```
me[0] + you[0] + 2^64*(overflow + me[1] + you[1]) + 2^128*(me[2]+you[2] + overflow1)
```
IOW if we have an overflow from `me[0] + you[0]` it needs to be added to `me[1] + you[1]` and we have an overflow if `overflow2` is true.

File: src/fast/u192.rs [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+pub struct U192(pub [u64; 3]);
+
+impl Mul for U192 {
+    type Output = Self;
+    fn mul(self, other: Self) -> Self {
+        // The generated implementation of this method was 360 lines long!
+        let me = &self.0;
+        let you = &other.0;
+
+        let mult = |m: usize, y: usize| unsafe {
```
Comment (lutter) [inline src/fast/u192.rs:21]: `usize` here is a bit strange - shouldn't this make it clear that we are multiplying two `u64` to get a `u128`?

File: src/fast/u192.rs [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+            transmute::<_, (u64, u64)>(me[m] as u128 * you[y] as u128)
```
Comment (lutter) [inline src/fast/u192.rs:22]: Doesn't that `transmute` assume something about the layout of a `u128`? The meaning of the first and second part of the result tuple will be different on little and big endian machines

File: src/fast/u192.rs [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+        let (low, hi0) = mult(1, 0);
+        let (r1, overflow0) = low.overflowing_add(r1);
+        let (low, hi1) = mult(0, 1);
+        let (r1, overflow1) = low.overflowing_add(r1);
+
+        let r2 = (hi0 + overflow0 as u64)
+            .wrapping_add(hi1 + overflow1 as u64)
+            .wrapping_add(me[2].wrapping_mul(you[0]))
+            .wrapping_add(me[1].wrapping_mul(you[1]))
+            .wrapping_add(me[0].wrapping_mul(you[2]));
```
Comment (lutter) [inline src/fast/u192.rs:35]: I haven't really gone through what's happening here; I'll do that tomorrow.

File: src/impls/bool.rs [hunk 13]
```diff
@@ -1,11 +1,12 @@
...
 
 impl StableHash for bool {
-    fn stable_hash<H: StableHasher>(&self, sequence_number: H::Seq, state: &mut H) {
+    #[inline]
+    fn stable_hash<H: StableHasher>(&self, field_address: H::Addr, state: &mut H) {
         profile_method!(stable_hash);
 
         if *self {
-            state.write(sequence_number, &[]);
+            state.write(field_address, &[]);
```
Comment (lutter) [inline src/impls/bool.rs:9]: This has consequences for how you can change struct definitions and still get the same hash; e.g., you can add a boolean field and as long as its value is `false`, you get the same hash as if the struct did not have the boolean field. Similar for empty strings and `None` options, but not for ints, where adding a field with value `0` still changes the hash.

It should be clearly documented (probably at the crate level or for `StableHasher`/`StableHash`) what changes to structs do not change the hash, and which ones do.

File: src/impls/vec.rs [hunk 14]
```diff
@@ -1,20 +1,22 @@
...
+    fn stable_hash<H: StableHasher>(&self, field_address: H::Addr, state: &mut H) {
         profile_method!(stable_hash);
 
-        for item in self.iter() {
-            item.stable_hash(sequence_number.next_child(), state);
+        for (index, item) in self.iter().enumerate() {
+            item.stable_hash(field_address.child(index as u64), state);
         }
-        self.len().stable_hash(sequence_number, state);
+        // This is needed to disambiguate when the last members are default
```
Comment (lutter) [inline src/impls/vec.rs:18]: I don't (fully) understand this comment - is this so we can distinguish, e.g., between a `Vec` with one entry `true`, and a `Vec` where the first entry is `true` followed by any number of `false`? Would be good if the comment spelled that out a little more (and the crate docs should mention that vecs of different length will have different hashes, even if they are filled with inert values)

File: src/lib.rs [hunk 15]
```diff
@@ -1,9 +1,49 @@
...
+
+/// Like Hash, but consistent across:
+/// * builds (independent of rustc version or std implementation details)
+/// * platforms (eg: 32 bit & 64 bit, x68 and ARM)
+/// * processes (multiple runs of the same program)
+pub trait StableHash {
+    fn stable_hash<H: StableHasher>(&self, field_address: H::Addr, state: &mut H);
+}
+
+pub trait FieldAddress: Sized {
```
Comment (lutter) [inline src/lib.rs:93]: This is a really important trait, and it should have comments on how to use it properly, and what `root` and `child` are used for; I don't understand at all what `unordered` is for, and how it should be used. I could definitely not implement this trait in a coherent way without such docs.

File: src/crypto/hasher.rs (hunk 1 above)
Comment (That3Percent) [reply src/crypto/hasher.rs:15]: Possibly, though it will no longer be relevant to graph-node. So, it's not a priority for me but it someone ever decides to use the crypto version for some reason in the future the option will be open to them.

File: src/crypto/hasher.rs (hunk 2 above)
Comment (That3Percent) [reply src/crypto/hasher.rs:18]: I got it from Jackson, who I think pulled it from here: [https://crypto.stackexchange.com/a/54546](https://crypto.stackexchange.com/a/54546)

File: src/crypto/hasher.rs (hunk 3 above)
Comment (That3Percent) [reply src/crypto/hasher.rs:25]: Correct!

File: src/crypto/hasher.rs (hunk 4 above)
Comment (That3Percent) [reply src/crypto/hasher.rs:28]: This comment is not speaking about code changes, but about hashing the struct incrementally and/or in parallel (which graph-node takes advantage of). It's saying that if you do such shenanigans, be sure to have struct.field have the same field address across runs. This comment was more salient for the previous API, which was `fieldAddress.nextChild()` instead of the current API which uses `fieldAddress.child(index)` which is much harder to screw up for a parallel implementation.

Your assertion about keeping the addresses fixed across code changes is also correct and important, just not what this comment is speaking to.

File: src/crypto/hasher.rs (hunk 5 above)
Comment (That3Percent) [reply src/crypto/hasher.rs:33]: Graph-node will not be relying on the cryptographic stable hash after this, so this TODO is still relevant to the library in the abstract as a potential performance improvement but not a priority for me or anyone else at E&N.

File: src/crypto/hasher.rs (hunk 6 above)
Comment (That3Percent) [reply src/crypto/hasher.rs:51]: This is an excellent suggestion. Unfortunately, I tried it in a branch and the result was slower. I didn't look much further into why at the time though, there's possibly still something there to speed this up. Repeat previous comment though about the crypto version no longer being relevant to graph-node.

File: src/crypto/hasher.rs (hunk 7 above)
Comment (That3Percent) [reply src/crypto/hasher.rs:100]: The comment is trying to be opaque about the format (only values returned by `to_bytes` are valid). I could (should) add an assertion for example that the value is less than the prime or something. The 257 comes from the fact that the prime is that long so it's not possible for a value that comes out of `to_bytes` to be longer.

File: src/fast/address.rs (hunk 8 above)
Comment (That3Percent) [reply src/fast/address.rs]: Ah, yes. This comment was written when this was `u64` instead of `u128`. That only requires ~4B values for a probability of 1/2, but the probability for 700M isn't so low that I would have been _shocked_ to find a collision at the time, it's just less than 1/2. I'll fix this comment.

File: src/fast/u192.rs (hunk 9 above)
Comment (That3Percent) [reply src/fast/u192.rs]: I think as-is is correct, but don't give up the fight unless you are totally convinced because it's important to get it right.

The reason we don't have an overflow if `overflow2` is true is because both `overflow1` and `overflow2` are tracking overflow of `r1`. One is from the possibility of overflow in the contribution of `me[1] + you[1]` to `r1` and one is from `r1 + overflow` to `r1`. These would be better named `overflow1a` and `overflow1b` and there is no `overflow2` since the contributions to `r2` use `wrapping_add`.

Comment (That3Percent) [reply src/fast/u192.rs]: I'm going to rename these to `overflow0`, `overflow1a` and `overflow1b`. If we did need to track overflow for `r2` it would be `overflow2a ... overflow2c`.

File: src/fast/u192.rs (hunk 10 above)
Comment (That3Percent) [reply src/fast/u192.rs:21]: The `usize` is because `mult(0, 1)` expands to `transmute::<_, (u64, u64)>(me[0] as u128 * you[1] as u128)`. Arguments are indexes to `me` and `you`... it just makes it more terse. I could make this `mult(me[0], you[1])` if you think that is better / more readable.

File: src/fast/u192.rs (hunk 11 above)
Comment (That3Percent) [reply src/fast/u192.rs:22]: Yes, I got lazy here. The implementation is supposed to produce the same result across architectures but does not. I think I'll just ban big-endian machines by panicking somewhere for now since I cannot actually test with one and there is very little chance of an Indexer using that for graph-node.

File: src/fast/u192.rs (hunk 12 above)
Comment (That3Percent) [reply src/fast/u192.rs:35]: Reminder: double check this math.

I think it's likely to be correct because I ran with parallel implementations of the generated `U192` across all the tests, including the stress test / benchmark, and never found any discrepancies. It would panic on any intermediate value not matching. Even so, it would be good to get a pair of eyes to make sure there are no rare edge cases missed here.

File: src/impls/bool.rs (hunk 13 above)
Comment (That3Percent) [reply src/impls/bool.rs:9]: For ints adding a value with `0` does not change the hash. No default values change the hash by design, including `vec![]`, etc. This is to support backward compatibility so we can add features to graph-node without changing PoIs. One weird case is that `Option(0)` _does_ change the hash but that's not possible to fix without creating a collision with `None`.

Why do you think adding a field with `0` changes the hash?

I will add some documentation to the crate, good suggestion.

File: src/impls/vec.rs (hunk 14 above)
Comment (That3Percent) [reply src/impls/vec.rs:18]: You are correct that this disambiguates between eg: `vec![true, false]` and `vec![true, false, false]` or `vec![true]`. (The test pointed to in the "See also" is an example of this, but using ints). It does not seem particularly "surprising" enough to be called out that vecs of different length will have different hashes even if filled with inert values _All_ possible values for _any_ possible schema do not collide within the same schema. That is, when holding the schema constant, the encoding is injective. (If you find this not to be true somewhere it's a show-stopping bug)

File: src/lib.rs (hunk 15 above)
Comment (That3Percent) [reply src/lib.rs:93]: I had not really intended this to be re-implemented, but I see your point. (Commenting is lacking across the board, really). Added lengthy comment.

**Comment by That3Percent** [comment]
@lutter All points addressed. The only changes pushed are comments, and banning big endian machines with a todo.

This was a great review and I'm very thankful for the thoughtful effort you put into it.

File: src/fast/u192.rs (hunk 10 above)
Comment (lutter) [reply src/fast/u192.rs:21]: No, as is is perfectly fine. I think I just tripped over the notation somehow

File: src/fast/u192.rs (hunk 11 above)
Comment (lutter) [reply src/fast/u192.rs:22]: I agree, a panic sounds like the most practical way to address this

File: src/fast/u192.rs (hunk 12 above)
Comment (lutter) [reply src/fast/u192.rs:35]: I just went through this, and AFAICT, this is correct.

File: src/fast/u192.rs (hunk 9 above)
Comment (lutter) [reply src/fast/u192.rs]: You are right - this is correct; I was getting confused about the different `overflow` variables

File: src/lib.rs (hunk 15 above)
Comment (lutter) [reply src/lib.rs:93]: Beautiful. This explanation helps readability a lot!

**Review by lutter — APPROVED** [review]
Looking forward to having this in `graph-node`!
